Add DSR IPv4 for NodePort BPF

brb opened the pull request:
This PR adds a direct server return (DSR) support for the NodePort BPF for IPv4 and in the direct routing mode.

The main idea of DSR is to avoid SNAT'ing an original request sent to an LB, so that a backend could directly reply to a client (the originator of the request) and the original source IP could be preserved.

To achieve this, we introduce a new IPv4 option which stores a NodePort service IP and port number. The option is set by `bpf_netdev` running on a public iface of an intermediate node which received the original request. Once the option has been set, the request (the dst IP addr of the request is DNAT'd to the backend IP addr) is forwarded to a node running the backend. After receiving the fwd'd request, `bpf_lxc` of the backend parses the option, stores the svc addr and port in the NAT table and sets the `dsr` bit in a CT entry.

When sending a reply to the client, `bpf_lxc` of the backend finds out that the `dsr` bit was set, does a lookup in the NAT table to find the mapping, and finally rewrites the source addr and port to the svc addr and port.

The current approach has a shortcoming that if the request size is `> (MTU - 8bytes)`, the request will be dropped after we append the IPv4 option.

To partially solve this, in the case of TCP we set the option only for `SYN` packets which should have an empty payload. However, the problem still exists for TCP with SYN cookies and UDP packets. For those cases, a client needs to decrease its MTU by 8bytes.

```release-note
Add direct server return (DSR) for NodePort BPF
```

DSR for IPv6 and docs will be submitted in a follow-up PR.

Reviewable per commit.

Related #8979 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9473)
<!-- Reviewable:end -->

brb committed:
daemon: Add --enable-dsr param

The param is used to enable direct server return (DSR) for
the NodePort BPF.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
helm: Add global.nodePort.dsr

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Add DSR for NodePort BPF IPv4

This commit adds a direct server return (DSR) support for the NodePort
BPF for IPv4 and in the direct routing mode.

The main idea of DSR is to avoid SNAT'ing an original request sent to an
LB, so that a backend could directly reply to a client (the originator
of the request) and the original source IP could be preserved.

To achieve this, we introduce a new IPv4 option which stores a NodePort
service IP and port number. The option is set by bpf_netdev running on
a public iface of an intermediate node which received the original request.
Once the option has been set, the request (the dst IP addr of the request is
DNAT'd to the backend IP addr) is forwarded to a node running the backend.
After receiving the fwd'd request, bpf_lxc of the backend parses the
option, stores the svc addr:port in the NAT table and sets the "dsr" bit
in a CT entry.

When sending a reply to the client, bpf_lxc finds out that the "dsr" bit
was set, does a lookup in the NAT table to find the mapping, and finally
rewrites the source addr and port to the svc addr and port.

The current approach has a shortcoming that if the request size is > (MTU -
8bytes), the request will be dropped after we append the IPv4 option.

To partially solve this, in the case of TCP we set the option only for
SYN packets which should have an empty payload. However, the problem
still exists for TCP with SYN cookies and UDP packets. For those cases,
a client needs to decrease its MTU by 8bytes.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Add DSR IPv4 integration tests

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Schedule test-nodeport-local pod on k8s2

For the DSR test case, we need to schedule the test-k8s2 (prev.
test-k8s1) pod on k8s2. Otherwise, a request from the
client-from-outside Docker container running on k8s1 to the pod
via k8s2 (sending via k8s1 does not test the DSR) would be dropped
by the kernel due to a routing loop detection mechanism:

1) k8s2 recv: client-from-outside (192.168.10.10) @ k8s1 -> k8s2:NodePort
2) k8s2 fwd to k8s1: client-from-outside (192.168.10.10) @ k8s1 -> Pod @ k8s1
3) k8s1 recv the packet on enp0s8, and has a route
   "192.168.10.0/24 dev $DOCKER_BRIDGE" <- kernel detects a potential loop.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -1048,9 +1061,39 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, __u8 *reason, _
 
 	if (ret == CT_NEW) {
+#ifdef ENABLE_DSR
+		if (ip4->ihl == 0x7) {
```
Please write a comment specifying this magic number.

aanm commented on bpf/lib/nodeport.h:
```diff
@@ -522,6 +522,87 @@ static __always_inline int nodeport_nat_ipv4_rev(struct __sk_buff *skb,
+	}
+
+	ip4->ihl += 0x2; // To accommodate u64 option
+	ip4->tot_len = bpf_htons(bpf_ntohs(ip4->tot_len) + 0x8);
```
the `0x2` is to accommodate u64 option, and what is the `0x8` for?

brb replied:
`ihl` counts in 32-bit words and stores IPv4 header len. `tot_len` counts in bytes and stores IPv4 header + header len.

aanm commented on test/k8sT/Services.go:
```diff
@@ -445,8 +445,27 @@ var _ = Describe("K8sServicesTest", func() {
+				url = getURL(helpers.K8s2Ip, data.Spec.Ports[0].NodePort)
+				doRequestsFromOutsideClient(url, 10, true)
+			})
+		})
```
How is this test checking the traffic was directly sent from the destination node, i.e. the backend, to the client?

brb replied:
It doesn't. I don't see any reasonably non-complex way to test this.

aanm reviewed (changes requested):
```
The current approach has a shortcoming that if the request size is > (MTU -
8bytes), the request will be dropped after we append the IPv4 option.

To partially solve this, in the case of TCP we set the option only for
SYN packets which should have an empty payload. However, the problem
still exists for TCP with SYN cookies and UDP packets. For those cases,
a client needs to decrease its MTU by 8bytes.
```

Why don't we adjust the MTU automatically of the public iface if the DSR option is set?

brb committed:
nat: GC NAT entries created by DSR

In the case of DSR, the following CT and NAT entries are created on
a host which runs a service endpoint and to which a client request is
forwarded:

* NAT: endpoint -> client (XLATE_SRC aka TUPLE_F_OUT)
* CT:  client -> endpoint (TUPLE_F_IN)

Previously, the CT GC ignored NAT entries when a corresponding CT entry was
of the TUPLE_F_IN type. Therefore, the DSR NAT entries could not have been
collected.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Allow provisioning arbitrary number of VMs

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Run k8s1-1.11 net-next on 3 VMs

This is going to be needed by some k8sT/Services.go tests.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Run curl from k8s3 instead of client-from-outside container

Previously, we ran curl from the "client-from-outside" container
in the tests which required sending requests from a third host.

We simulated the third host by running a container
("client-from-outside") in a Docker network which was not managed
by Cilium.

Unfortunately, requests sent to a NodePort service from the container
were handled by bpf_sock.c which prevented from testing the NodePort
implementation in bpf_netdev.c.

Fix it by introducing a "real" host, and run curl from it.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Improve indentation of kubeproxy-free tests

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Fix DatapathConfiguration tests when running on >2 nodes

Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -52,6 +52,7 @@ cilium-agent [flags]
       --disable-endpoint-crd                                  Disable use of CiliumEndpoint CRD
       --disable-k8s-services                                  Disable east-west K8s load balancing by cilium
       --egress-masquerade-interfaces string                   Limit egress masquerading to interface selector
+      --enable-dsr                                            Enable direct server return NodePort BPF (beta)
```
Should we do `--node-port-mode=dsr`, so possible values could be {dsr,xdp,legacy} then also used in combination, e.g. `--node-port-mode=dsr,xdp`. Thoughts?

brb replied:
SGTM, just maybe `s/legacy/snat`.

borkmann replied:
Yes, `snat` sounds good to me.

borkmann replied:
With `legacy` I actually meant for older kernels that do not have host reachable service support.

borkmann commented on daemon/daemon_main.go:
```diff
@@ -1085,6 +1088,17 @@ func initEnv(cmd *cobra.Command) {
+		case !option.Config.EnableNodePort:
+			log.Fatal("DSR can be enabled only for NodePort BPF (--enable-node-port)")
+		case option.Config.EnableDSR && option.Config.EnableIPv6:
+			log.Fatal("DSR cannot be used with IPv6")
```
PR for IPv6 support will also come before 1.7 freeze, right?

brb replied:
Yep, after this PR has been merged.

borkmann commented on daemon/daemon_main.go:
```diff
@@ -1085,6 +1088,17 @@ func initEnv(cmd *cobra.Command) {
+		case option.Config.EnableDSR && option.Config.EnableIPv6:
+			log.Fatal("DSR cannot be used with IPv6")
+		case option.Config.Tunnel != option.TunnelDisabled:
+			log.Fatal("DSR cannot be used with tunnel")
```
Why is tunneling not supported? Couldn't the backend do the same logic here as in, sending the reply directly back (not via tunnel dev but phys dev)?

brb replied:
Discussed offline: I will do it as a follow-up (v1.8).

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -554,6 +554,22 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
+			struct ipv4_nat_entry *entry = snat_v4_lookup(&nat_tup);
+			if (entry) {
+				int ret1 = snat_v4_rewrite_egress(skb, &nat_tup,
+								  entry, l4_off);
```
nit: could we just reuse `ret`? Also, I'd stick with current style and move declaration of `struct ipv4_nat_entry *entry` up, then leave a newline before the nat_tup overrides.

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -1048,9 +1065,41 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, __u8 *reason, _
+#ifdef ENABLE_DSR
+		// Check whether IPv4 header contains a 64-bit option (IPv4 header
+		// w/o option (5 x 32-bit words) + the DSR option (2 x 32-bit words))
+		if (ip4->ihl == 0x7) {
```
What happens if the IPv4 packet carries already other options aside our custom one? Why ignore DSR in that case (seems unexpected)? At min needs to be documented that this would be incompatible with CIPSO / others.

borkmann replied:
Note that this will turn off GRO at the server side. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ipv4/af_inet.c#n1435 . We should make sure that it's still faster than SNAT.

brb replied:
> What happens if the IPv4 packet carries already other options aside our custom one

I didn't want to increase code complexity in terms of the verifier by adding a proper IPv4 options parser. I will document this, and implement as a follow-up.

> Note that this will turn off GRO at the server side

Good point, thanks for noting.

borkmann commented on bpf/bpf_netdev.c:
```diff
@@ -714,14 +714,14 @@ int to_netdev(struct __sk_buff *skb)
-#if defined(ENABLE_NODEPORT) || defined(ENABLE_MASQUERADE)
-#ifdef ENABLE_NODEPORT
+#if (defined(ENABLE_NODEPORT) && !defined(ENABLE_DSR)) || defined(ENABLE_MASQUERADE)
+# if defined(ENABLE_NODEPORT) && !defined(ENABLE_DSR)
```
Did ENABLE_MASQUERADE drop?

brb replied:
Noup, it's still present at the line 717.

borkmann commented on bpf/lib/l3.h:
```diff
@@ -174,4 +174,34 @@ static inline __u8 __inline__ get_min_encrypt_key(__u8 peer_key)
+		csum += *h++;
+
+       iph->check = ~((csum & 0xffff) + (csum >> 16));
+}
```
csum_diff() helper?

brb committed:
daemon: Replace --enable-dsr with --node-port-mode

As we are planning to support multiple mutually inclusive modes for
NodePort, introduce a flag to store them. Also, re-use the flag
for enabling the DSR option.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on test/helpers/kubectl.go:
```diff
@@ -242,6 +242,7 @@ func CreateKubectl(vmName string, log *logrus.Entry) (k *Kubectl) {
 func (kub *Kubectl) LabelNodes() {
 	kub.ExecMiddle(fmt.Sprintf("%s label --overwrite node k8s1 cilium.io/ci-node=k8s1", KubectlCmd))
 	kub.ExecMiddle(fmt.Sprintf("%s label --overwrite node k8s2 cilium.io/ci-node=k8s2", KubectlCmd))
+	kub.ExecMiddle(fmt.Sprintf("%s label --overwrite node k8s3 cilium.io/ci-node=k8s3", KubectlCmd))
```
This command will fail, can you detect the number of nodes running and avoid running this command if it's not running node 3?

brb replied:
Sure, but we just ignore a returned error. I will do it as a follow-up.

aanm replied:
Please create a GH issue so we can track this.

brb replied:
Pushed the change which fixes this.

aanm commented on test/k8sT/manifests/demo_ds.yaml:
```diff
@@ -13,6 +13,17 @@ spec:
       labels:
         zgroup: testDS
     spec:
+      # Some tests (e.g. DSR) requires to send requests to testds from a host
+      # which does not run any of its pods. We've chosen k8s3 to be such host.
+      affinity:
```
This file is used for many tests. I would strongly suggest to have a new file dedicated for the tests that can only be run with +3 VMs. Can be done in a follow up PR.

brb replied:
Will do it as a follow up!

aanm replied:
Please create a GH issue so we can track this.

brb replied:
Actually just did in this PR.

aanm reviewed (changes requested):
Also needs documentation WRT. can be done in a follow up PR

borkmann reviewed (approved):
Looks okay, my preference would be to move more of the bpf_lxc DSR code into the nodeport header. We should alos properly prefix the set_dsr_opt4() and other functions sitting in the nodeport header.

Things still missing till final release:

- IPv6 support
- Tunnel support (IPv4/IPv6)
- Switch to IPIP instead of IP options to avoid breaking GRO
- Documentation

borkmann commented on bpf/lib/nat.h:
```diff
@@ -441,6 +441,52 @@ static __always_inline bool snat_v4_can_skip(const struct ipv4_nat_target *targe
+	off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
+	switch (tuple.nexthdr) {
+	case IPPROTO_TCP:
+	case IPPROTO_UDP:
```
What about ICMP which is handled partially in NAT engine?

brb replied:
An ICMP packet cannot be sent to a NodePort service. I will add a comment.

borkmann commented on bpf/lib/nat.h:
```diff
@@ -441,6 +441,52 @@ static __always_inline bool snat_v4_can_skip(const struct ipv4_nat_target *targe
+
+	int ret = map_update_elem(&SNAT_MAPPING_IPV4, &tuple, &state, 0);
+	if (!ret)
+		return ret;
```
You meant `if (ret)` ?

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -554,6 +554,24 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 			return DROP_MISSED_TAIL_CALL;
 		}
+# ifdef ENABLE_DSR
+		if (ct_state.dsr) {
```
Do we have a case for DSR outside of nodeport? Maybe this should be moved under `ct_state.node_port` instead to avoid two extra tests on bitfields in fast-path when not hit?

brb replied:
Discussed offline: we might support enabling of DSR for some not all services (via k8s annotations), so keeping as it is for now.

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -554,6 +554,24 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
+								  entry, l4_off);
+				if (ret != 0) {
+					return ret;
+				}
```
Should likely be an inline helper for nodeport.. same inside ipv4_policy.

brb replied:
I will do it as a follow-up.

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -1048,9 +1067,41 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, __u8 *reason, _
+				return DROP_INVALID;
+			opt1 = bpf_ntohl(opt1);
+
+			if ((opt1 & 0xffff0000) == DSR_IPV4_OPT_32) {
```
Magic num, needs define.

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -500,6 +499,98 @@ static __always_inline int nodeport_nat_ipv4_rev(struct __sk_buff *skb,
+	if (l3_csum_replace(skb, ETH_HLEN + offsetof(struct iphdr, check),
+	    0, sum, 0) < 0) {
+		return DROP_CSUM_L3;
+	}
```
Did you check what happens under GSO/TSO? Meaning, large, GRO-aggregated train of skbs comes in, creates a bit one, we adjust room and expect GSO/TSO later on to split it again. Does this work and takes these options into account (I'm currently unsure, but worth double checking the GSO engine code and to run some experiments)?

brb replied:
Discussed offline: as we set the option only for the first packet, the GRO and GSO/TSO implications are not worrisome.

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -637,6 +729,7 @@ static inline int nodeport_lb4(struct __sk_buff *skb, __u32 src_identity)
 		else
 			return ret;
 	}
+	ct_state_new.orig_dport = key.dport;
```
Did this slip in or why is it needed? (short comment would be nice)

brb replied:
A leftover from the rebase.

brb committed:
datapath: Fix 4.19.57 verifier complaint for bpf_lxc w/ DSR

Add revalidate_data() to avoid verifier complaining on the 4.19.57
kernel when loading bpf_lxc.o:

level=warning msg=" R0=inv(id=0,umax_value=2147483647,var_off=(0x0; 0x7fffffff)) R1=inv0 R3=inv0 R4=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R5=inv(id=0) R6=ctx(id=0,off=0,imm=0) R7=inv0 R8=inv(id=0) R9=inv0 R10=fp0,call_-1 fp-104=0 fp-112=0 fp-120=0 fp-152=0 fp-216=0" subsys=datapath-loader
level=warning msg="1007: (7b) *(u64 *)(r10 -152) = r3" subsys=datapath-loader
level=warning msg="1008: (b7) r7 = 0" subsys=datapath-loader
level=warning msg="1009: (71) r1 = *(u8 *)(r8 +0)" subsys=datapath-loader
level=warning msg="R8 invalid mem access 'inv'" subsys=datapath-loader

Signed-off-by: Martynas Pumputis <m@lambda.lt>
